crypto/hpke/hpke.c

```diff
@@ -1833,6 +1854,190 @@ int OSSL_HPKE_CTX_export(OSSL_HPKE_CTX *ctx,
+        if (ctx->info == NULL) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
+            return 0;
+        }
+        memcpy(ctx->info, info, infolen);
+        ctx->infolen = infolen;
+    }
+    return 1;
+}
+
+/**
+ * @brief recipient decapsulation function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param recippriv is the EVP_PKEY form of recipient private value
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * Following this, OSSL_HPKE_CTX_export can be called.
+ */
+int OSSL_HPKE_decap(OSSL_HPKE_CTX *ctx,
+                    const unsigned char *enc, size_t enclen,
+                    EVP_PKEY *recippriv,
+                    const unsigned char *info, size_t infolen)
+{
+    int erv = 1;
+
+    if (ctx == NULL || enc == NULL || enclen == 0 || recippriv == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (ctx->shared_secret != NULL) {
+        /* only allow one encap per OSSL_HPKE_CTX */
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+    erv = hpke_decap(ctx, enc, enclen, recippriv);
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (info != NULL) {
+        if (infolen == 0) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+            return 0;
+        }
+        ctx->info = OPENSSL_malloc(infolen);
+        if (ctx->info == NULL) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE);
+            return 0;
+        }
+        memcpy(ctx->info, info, infolen);
+        ctx->infolen = infolen;
+    }
+    return 1;
+}
+
+/**
+ * @brief new sender seal function
+ * @param ctx is the pointer for the HPKE context
+ * @param ct is the ciphertext output
+ * @param ctlen is the size the above
+ * @param aad is the aad parameter
+ * @param aadlen is the size the above
+ * @param pt is the plaintext
+ * @param ptlen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * This can be called multiple times
+ */
+int OSSL_HPKE_seal(OSSL_HPKE_CTX *ctx,
+                   unsigned char *ct, size_t *ctlen,
+                   const unsigned char *aad, size_t aadlen,
+                   const unsigned char *pt, size_t ptlen)
+{
+    int erv = 1;
+
+    if (ctx == NULL || ct == NULL || ctlen == NULL || *ctlen == 0
+        || pt == NULL || ptlen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    if (ctx->shared_secret == NULL) {
+        /* need to have done an encap first, info can be NULL */
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+    erv = hpke_do_rest(ctx, OSSL_HPKE_OP_ENC, ct, ctlen,
```

Comment:
I still don't agree with what this is doing :) i.e.
 - hpke_do_rest() should be renamed to hpke_schedule() and just do the schedule.. It would cache the generated nonce and key in the ctx.
- OSSL_HPKE_encap and decap call the hpke_schedule()
- OSSL_HPKE_seal/open only do the final bit that you have in hpke_do_rest() currently.

Reply:
It looks like this is at least a internal problem now that the API has changed.. i.e. The API wont need to change with either method..

Reply:
overtaken by events